lib/Transforms/IPO/StructAnalysisTransformAnalyzer.cpp

```diff
@@ -0,0 +1,865 @@
+                         cl::Hidden,
+                         cl::desc("Set COLD_RATIO in struct split algorithm"));
+
+static cl::opt<unsigned> StructSplitDistanceThreshold(
+    "struct-analysis-split-distance-threshold", cl::init(120), cl::Hidden,
+    cl::desc("Set DISTANCE_THRESHOLD in struct split algorithm"));
+
+static cl::opt<unsigned>
+    StructSplitMaxSize("struct-analysis-split-max-size", cl::init(32),
+                       cl::Hidden,
+                       cl::desc("Set MAX_SIZE in struct split algorithm"));
+
+static cl::opt<unsigned> StructSplitSizePenalty(
+    "struct-analysis-split-size-penalty", cl::init(20), cl::Hidden,
+    cl::desc("Set SIZE_PENALTY in struct split algorithm"));
+
+static cl::opt<unsigned> StructSplitMaxThresholdUpdates(
+    "struct-analysis-split-max-threshold-updates", cl::init(3), cl::Hidden,
+    cl::desc(
+        "Max number of lowering threshold allowed when making suggestions"));
+
+// Functions of StructTransformAnalyzer
+// Constructor of base class
+StructTransformAnalyzer::StructTransformAnalyzer(
+    const Module &CM, const StructType *ST, const CloseProximityBuilder *CPB,
+    const DICompositeType *DI)
+    : StructureType(ST), CPBuilder(CPB),
+      NumElements(StructureType->getNumElements()), DebugInfo(DI),
+      Eligibility(true) {
+  if (NumElements <= 2) {
+    Eligibility = false;
+  }
+  // Calculate field sizes and establish a set of all fields considered to
+  // transform
+  FieldSizes.resize(NumElements);
+  for (unsigned i = 0; i < NumElements; i++) {
+    FieldSizes[i] =
+        CM.getDataLayout().getTypeSizeInBits(StructureType->getElementType(i)) /
+        8;
+    FieldsToTransform.insert(i);
+  }
+
+  // Check if any field could be a padding
+  mapFieldsToDefinition();
+  for (unsigned i = 0; i < NumElements; i++) {
+    if (FieldDI[i] == nullptr) {
+      outs() << "Field " << i + 1
+             << " is actually a padding, ignore in suggestions...\n";
+      FieldsToTransform.erase(i);
+    }
+  }
+}
+
+// Functions that maps LLVM struct types to struct definition in source
+// code. The function constructs an array FieldDI, where each index represents
+// each field in LLVM struct type and the content in each index is the debug
+// info (FieldDebugInfo type) that contains info like name of the field,
+// type of the field and actual field number (in case of padding). If a field
+// is a padding, we save a nullptr to the according index.
+// When DebugInfo of the struct is not retrieved (which is the
+// default case for now), we use some heuristics to guess which field
+// might be a padding. Other elements in FieldDI stores FieldDebugInfo type
+// that only contains a field number in original definition (excluding padding)
+void StructTransformAnalyzer::mapFieldsToDefinition() {
+  FieldDI.resize(NumElements);
+  if (DebugInfo) {
+    // Not actually using this part because DebugInfo can't be correctly
```

Comment:
If it is never tested, remove this part and add a comment on how mone might handle DebugInfo.